mtt/src/mtt_types.rs

```diff
@@ -0,0 +1,929 @@
+        let l2_entry = MttL2Directory::entry_for_addr(addr, self.mtt_info.mtt_l2_base)?;
+        match l2_entry {
+            Confidential1G(_) => Ok(MttMemoryType::Confidential),
+            NonConfidential1G(_) => Ok(MttMemoryType::NonConfidential),
+            Mixed2M(mixed_2m_entry) => {
+                if mixed_2m_entry.is_conf(Aligned2MAddr::new(addr)?) {
+                    Ok(MttMemoryType::Confidential)
+                } else {
+                    Ok(MttMemoryType::NonConfidential)
+                }
+            }
+            L1Entry(l1_entry) => {
+                if l1_entry.is_conf(addr) {
+                    Ok(MttMemoryType::Confidential)
+                } else {
+                    Ok(MttMemoryType::NonConfidential)
+                }
+            }
+        }
+    }
+
+    fn allocate_l1_page(&mut self) -> Result<SupervisorPageAddr> {
+        let addr = self
+            .mtt_info
+            .l1_page_list
+            .peek()
+            .ok_or(InsufficientL1Pages)?;
+        if self.get_4k_region_type(addr)? != MttMemoryType::Confidential {
+            return Err(NonConfidentialL1Page);
+        }
+        // Unwrap Ok: Must succeed given check above
+        let addr = self.mtt_info.l1_page_list.pop().unwrap().addr();
+        Ok(addr)
+    }
+
+    fn set_4k_region_type(
+        &mut self,
+        addr: SupervisorPageAddr,
+        memory_type: MttMemoryType,
+    ) -> Result<bool> {
+        let is_conf = matches!(self.get_4k_region_type(addr)?, MttMemoryType::Confidential);
+        let mtt_needs_update = (is_conf && memory_type == MttMemoryType::NonConfidential)
+            || (!is_conf && memory_type == MttMemoryType::Confidential);
+
+        if mtt_needs_update {
+            let l2_entry = MttL2Directory::entry_for_addr(addr, self.mtt_info.mtt_l2_base)?;
+            let mut l1_entry = match l2_entry {
+                Confidential1G(_) | NonConfidential1G(_) => {
+                    let l1_page = self.allocate_l1_page()?;
+                    let aligned_1g = PageAddr::new(SupervisorPhysAddr::supervisor(
+                        PageSize::Size1G.round_down(addr.bits()),
+                    ))
+                    .ok_or(InvalidAddress)?;
+                    let aligned_1g_addr = Aligned1GAddr::new(aligned_1g)?;
+                    MttL2Directory::convert_1g_to_64m(aligned_1g_addr, self.mtt_info.mtt_l2_base)?;
+                    match MttL2Directory::entry_for_addr(addr, self.mtt_info.mtt_l2_base)? {
+                        Mixed2M(mixed_2m_entry) => {
+                            mixed_2m_entry.convert_to_l1_entry(l1_page, self.mtt_info.mtt_l2_base)
+                        }
+                        _ => unreachable!(),
+                    }
+                }
+                Mixed2M(mixed_2m_entry) => {
+                    let l1_page = self.allocate_l1_page()?;
+                    mixed_2m_entry.convert_to_l1_entry(l1_page, self.mtt_info.mtt_l2_base)
+                }
```

Comment:
So this logic for demoting from one L2 entry size to another gets repeated several times. It's a little tricky, so I think it's worthwhile to try to avoid the duplication.

How about a loop that looks (very roughly) like this for doing the conversion bewteen C <-> NC for a range of addresses?

```
    while current != end {
        match l2_dir.entry_for_addr(current) {
            Entry1G(entry) => {
                if entry.mem_type() == target_type {
                    current += min(end - current, end_of_1g_range - current);
                } else if current % 1G == 0 && end - current >= 1G {
                    entry.convert();
                    current += 1G;
                } else {
                    entry.split_to_64m();
                }                
            }
            Entry64M(entry) => {
                if entry.mem_type_for_2m_range(current) == target_type {
                    current += min(end - current, end_of_2m_range - current);
                } else if current % 2M == 0 && end - current >= 2M {
                    entry.convert_2m_range(current);
                    current += 2M;
                } else {
                    entry.convert_to_l1_table(get_l1_page);
                }
            }
            EntryL1Table(table) => {
                if entry.mem_type_for_4k_range(current) == target_type {
                    current += 4k;
                } else {
                    table.convert_4k_range(currnet);
                    current += 4k;
                }
            }
        }
    }
```

Reply:
Seems like this loop is identical to what the MttRange iterator is already doing?

Reply:
it isn't. notice how, as written, you have to do the 1g -> 64m demotion in two places (when changing a 2m range, and when changing a 4k range). what i'm suggesting is that we only have one place where these transition between L2 entry types happens. the key here is that the loop i had above does not advance to the next range whenever we do a demotion, and instead retries with the new L2 entry.